File: VB/Reporting_how-to-override-the-existing-behavior-when-dropping-an-item-from-the-field-list-in-e478/FieldListDragDropService.vb
Imports System
Imports System.ComponentModel.Design
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraReports.Design
Imports DevExpress.XtraReports.UI

Namespace FieldListDragDrop
	Public Class MyFieldListDragDropService
		Implements IFieldListDragDropService

		Private host As IDesignerHost
		Public Sub New(ByVal host As IDesignerHost)
			Me.host = host
		End Sub
		Public Function GetDragHandler() As IDragHandler Implements IFieldListDragDropService.GetDragHandler
			Return New MyFieldDragHandler(host)
		End Function
	End Class
	Public Class MyFieldDragHandler
		Inherits FieldDragHandler

		Public Sub New(ByVal host As IDesignerHost)
			MyBase.New(host)
		End Sub

		Public Overrides Sub HandleDragDrop(ByVal sender As Object, ByVal e As DragEventArgs)
			AdornerService.ResetSnapping()
			RulerService.HideShadows()

            Dim parent As XRControl = BandViewSvc.GetControlByScreenPoint(New System.Drawing.Point(e.X, e.Y))
			If parent Is Nothing Then
				Return
			End If

			Dim xRLabel As New XRRichText()
			Dim location As PointF = GetDragDropLocation(e, xRLabel, parent)

			DesignTool.AddToContainer(Host, xRLabel)

			xRLabel.LocationF = location
			xRLabel.Size = New Size(100, 25)
			xRLabel.DataBindings.Add("Rtf", Nothing, "test")
		End Sub

		Private Function GetDragDropLocation(ByVal e As DragEventArgs, ByVal control As XRControl, ByVal parent As XRControl) As PointF
			Dim bandPoint As PointF = EvalBandPoint(e, parent.Band)
			bandPoint = BandViewSvc.SnapBandPoint(bandPoint, parent.Band, control, New XRControl() { control })
			Dim screenPoint As PointF = BandViewSvc.ControlViewToScreen(bandPoint, parent.Band)
			Return BandViewSvc.ScreenToControl(New RectangleF(screenPoint, SizeF.Empty), parent).Location
		End Function

		Private Shared Function GetDragData(ByVal dataObject As IDataObject) As DragDataObject
			Return DirectCast(dataObject.GetData(GetType(DragDataObject)), DragDataObject)
		End Function
	End Class
End Namespace

File: CS/Reporting_how-to-override-the-existing-behavior-when-dropping-an-item-from-the-field-list-in-e478/FieldListDragDropService.cs
using System;
using System.ComponentModel.Design;
using System.Drawing;
using System.Windows.Forms;
using DevExpress.XtraReports.Design;
using DevExpress.XtraReports.UI;

namespace FieldListDragDrop {
    public class MyFieldListDragDropService : IFieldListDragDropService {
        IDesignerHost host;
        public MyFieldListDragDropService(IDesignerHost host) { this.host = host; }
        public IDragHandler GetDragHandler() {
            return new MyFieldDragHandler(host);
        }
    }
    public class MyFieldDragHandler : FieldDragHandler {
        public MyFieldDragHandler(IDesignerHost host)
            : base(host) {
        }
        public override void HandleDragDrop(object sender, DragEventArgs e) {
            AdornerService.ResetSnapping();
            RulerService.HideShadows();

            XRControl parent = BandViewSvc.GetControlByScreenPoint(new Point(e.X, e.Y));
            if (parent == null)
                return;

            XRRichText demoRichText = new XRRichText();
            PointF location = GetDragDropLocation(e, demoRichText, parent);

            DesignTool.AddToContainer(Host, demoRichText);

            demoRichText.LocationF = location;
            demoRichText.Size = new Size(100, 25);
            demoRichText.DataBindings.Add("Rtf", null, "test");
        }

        private PointF GetDragDropLocation(DragEventArgs e, XRControl control, XRControl parent) {
            PointF bandPoint = EvalBandPoint(e, parent.Band);
            bandPoint = BandViewSvc.SnapBandPoint(bandPoint, parent.Band, control, new XRControl[] { control });
            PointF screenPoint = BandViewSvc.ControlViewToScreen(bandPoint, parent.Band);
            return BandViewSvc.ScreenToControl(new RectangleF(screenPoint, SizeF.Empty), parent).Location;
        }

        static DragDataObject GetDragData(IDataObject dataObject) {
            return (DragDataObject)dataObject.GetData(typeof(DragDataObject));
        }
    }
}
